PubNubUnity/Assets/PubNub/PlayModeTests/PlayModeTests.cs

```diff
@@ -5761,5 +5752,151 @@ public IEnumerator TestMessageCounts()
 			pubnub.CleanUp();
 		}
 
+		[UnityTest]
+		public IEnumerator TestGrantToken() {
+			PNConfiguration pnConfiguration = PlayModeCommon.SetPAMPNConfig(false);
+			PubNub pnPAM = new PubNub(pnConfiguration);
+
+			System.Random r = new System.Random ();
+			pnConfiguration.UUID = "UnityTestConnectedUUID_" + r.Next (10000);
+			int ran = r.Next (10000);
+			int ran2 = r.Next (10000);
+			string userid = "userid"  + ran;
+			string spaceid = "spaceid"  + ran;
+			string userid2 = "userid"  + ran2;
+			string spaceid2 = "spaceid"  + ran2;
+			
+			var resChannels = new Dictionary<string, int>(){
+				{userid, 31},
+				{userid2, 239},
+			};
+
+			var resGroups = new Dictionary<string, int>(){
+				{spaceid, 103},
+				{spaceid2, 231},
+			};
+			var resUUIDs = new Dictionary<string, int>(){
+				{pnConfiguration.UUID, 215},
+			};
+
+			// publish expect 403
+			bool testReturn = false;
+			pnPAM.Publish().Channel(userid).Message("test 403").Async((result, status) => {
+				testReturn = status.Error.Equals(true);
+				Assert.True(status.StatusCode.Equals(403), status.StatusCode.ToString());
+			});
+			yield return new WaitForSeconds(PlayModeCommon.WaitTimeBetweenCalls2);
+			Assert.True(testReturn, "test didn't return or the response was not expected");
+			
+			string token = "";
+			pnPAM.GrantToken().SetParams(				
+				resChannels, 
+				resGroups,				
+				new Dictionary<string, int>(), 
+				new Dictionary<string, int>(), 
+				resUUIDs,
+				new Dictionary<string, int>(),
+				new Dictionary<string, int>(),				
+				new Dictionary<string, int>(), 
+				new Dictionary<string, int>(), 
+				new Dictionary<string, int>(),
+				3, 
+				pnConfiguration.UUID
+				)
+				.Async((result, status) => {
+				Debug.Log("GrantToken response:::" + result.Token);
+				token = result.Token;
+			});
+
+			yield return new WaitForSeconds (PlayModeCommon.WaitTimeBetweenCalls4);
+			pnPAM.SetToken(token);
+			
+			TokenContents p = pnPAM.ParseToken(token);
```

Comment:
for ParseToken, spec says "....including implementation details needed to check validity and expiration."
Want to know if validity and expiration checks are added.

Reply:
Do you mean "Returns the ttl, permissions and other information stored in the token including implementation details needed to check validity and expiration.", there are no validation for the returned object. The object which is returned is of the type TokenContents, which has he ttl, permissions and other information.

Reply:
I mean p.IsExpired() which returns bool
and p.IsValidToken() ..I am not sure what is the criteria for checking validity.
For IsExpired, we can check based on timestamp and ttl.

Reply:
I don't see any such properties for the returned object.

Reply:
I know there are no properties in TokenContents, but for the customer to validate, don't we provide that interface to check whether token is expired or not  AND token is valid or not.
Asking this based on the verbiage of the specs.

Reply:
need to check with the feature champion

Reply:
um...may be validation is not required if we check internally "Validation: must include at least one of resources or patterns"

Reply:
Also signature seems not translated to base64 string.